hashicorp/terraform
# Prevent duplicate plan output for TFE releases that don't support SRO for CLI

A customer reported seeing duplicate plan output when using Terraform v1.4.x against a TFE release that did not support enabling structured logging for runs created via the CLI. They see duplicate plan output because these releases will silently disable the setting when a run is created, resulting in plan output being written in human-readable format to the streamed log file. The only requirement to call the new JSON plan renderer (which generates plan output) was that the remote workspace needs to have the structured run output setting enabled. Therefore, plan output would be read from the streamed log file and subsequently re-generated by the plan renderer. 

In short, this fix makes the check "TFE-aware". The requirement to call the JSON plan renderer for a cloud backend configured against TFE is:
- If it is a knowable TFE version (i.e, TFE releases that include the `X-TFE-Version` header), then the release must be equal to or later than `202302-1`. This is the release version that adds SRO support for CLI-driven runs.
- If it is an unknowable version, we automatically know SRO will be silently disabled. This would indicate at the very least a version that predates the `X-TFE-Version` change. 

**NOTE**: The check still requires structured run output to be enabled on a TFE workspace for the renderer to be called.



## Target Release



1.4.5

## Draft CHANGELOG entry



###  BUG FIXES

 

-  `cloud`: Fixes duplicate plan output for TFE releases that don't support enabling structured logging for CLI runs.

## Testing
In order to smoke test this, you will need a TFE release older than `202302-1`. 
1. Make sure you're using a release version of Terraform v1.4, `tfenv install 1.4.0 && tfenv use 1.4.0`
1. Create any config. Make sure you've configured the `cloud` backend against the TFE installation. And `terraform plan` (your workspace, if created on `terraform init`, should have SRO enabled by default). 
1. You should see duplicate plan output.
1. Great! Now checkout this branch locally and `go install`. Run a plan using the installed binary: `$GOPATH/bin/terraform plan`
1. You should no longer see duplicate plan output. 
1. Feel free to test this against a release of TFE that supports SRO for CLI runs or TFC. Both cases should not yield duplicate plan output.
